### DISCUSSION TITLE
单 Epoll 多线程 IO 模型 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by giscus** [SF-Zhou/sf-zhou.github.io#195] [Announcements]
# 单 Epoll 多线程 IO 模型 | SF-Zhou's Blog

https://sf-zhou.github.io/linux/single_epoll_multi_workers.html

**Comment by SF-Zhou** [comment]
笔者按照文中的思路实现代码，然后做压测，Echo 的 QPS 可以跑到 400w/s，server 侧 CPU 4200%，单核心贡献 QPS 近 10w/s。余下的瓶颈主要是在 `folly::CPUThreadPoolExecutor` 中的队列，后面有机会替换成带 Work-stealing 的线程池看看。

**Comment by SF-Zhou** [comment]
最近尝试使用 io_uring 优化掉现有存储系统中的 libaio，感觉使用 io_uring 完成文中的读写操作要比线程池启动任务处理更好，有空我测试下。
